modin/core/io/column_stores/parquet_dispatcher.py

```diff
@@ -557,15 +557,10 @@ def build_query_compiler(cls, dataset, columns, index_columns, **kwargs):
         )
         index, sync_index = cls.build_index(dataset, partition_ids, index_columns)
         remote_parts = cls.build_partition(partition_ids, column_widths)
-        if len(partition_ids) > 0:
-            row_lengths = [part.length() for part in remote_parts.T[0]]
```

Comment:
It can be calculated later but there might be slight overhead in doing so. The proposal from @YarShev was to have `build_index` return the lengths since we already have to materialize the index. I think we should do something like that instead.

Reply:
@pyrito we can change the metadata getters to avoid overhead in this and many other cases. what do you think?
```bash
diff --git a/modin/core/dataframe/pandas/dataframe/dataframe.py b/modin/core/dataframe/pandas/dataframe/dataframe.py
index 4bd3b538..36801541 100644
--- a/modin/core/dataframe/pandas/dataframe/dataframe.py
+++ b/modin/core/dataframe/pandas/dataframe/dataframe.py
@@ -247,12 +247,15 @@ class PandasDataframe(ClassLogger):
         """
         if self._row_lengths_cache is None:
             if len(self._partitions) > 0:
-                (
-                    index,
-                    self._row_lengths_cache,
-                ) = self._compute_axis_labels_and_lengths(0)
-                if self._index_cache is None:
-                    self._index_cache = index
+                row_parts = self._partitions.T[0]
+                if self._index_cache is not None:
+                    # do not do extra work to get an index that is already known
+                    self._row_lengths_cache = [part.length() for part in row_parts]
+                else:
+                    (
+                        self._index_cache,
+                        self._row_lengths_cache,
+                    ) = self._compute_axis_labels_and_lengths(0)
             else:
                 self._row_lengths_cache = []
         return self._row_lengths_cache
@@ -269,12 +272,15 @@ class PandasDataframe(ClassLogger):
         """
         if self._column_widths_cache is None:
             if len(self._partitions) > 0:
-                (
-                    columns,
-                    self._column_widths_cache,
-                ) = self._compute_axis_labels_and_lengths(1)
-                if self._columns_cache is None:
-                    self._columns_cache = columns
+                col_parts = self._partitions[0]
+                if self._columns_cache is not None:
+                    # do not do extra work to get columns that is already known
+                    self._column_widths_cache = [part.width() for part in col_parts]
+                else:
+                    (
+                        self._columns_cache,
+                        self._column_widths_cache,
+                    ) = self._compute_axis_labels_and_lengths(1)
             else:
                 self._column_widths_cache = []
         return self._column_widths_cache

```

Reply:
@anmyachev I'm not sure I fully understand. Is the proposal to try to do all the calculations lazy? In other words, don't calculate the row or column lengths here?

Reply:
No, my suggestion is not to return indices from remote functions (on the basis of which lengths are calculated), but immediately the length (just a number). This greatly reduces the time to serialize / deserialize the index in some cases (for example, a multi-index). I do not have specific numbers, but it seems to me that the performance gain here should be obvious.

Reply:
@anmyachev Oh I see. To be honest, I'm not completely sure. My suggestion would be to try it yourself and see if you are getting good perf/correctness from not returning the index and building it lazily. I'm not sure if that would run into any issues.

Reply:
@pyrito my suggestion for now is to lazily calculate lengths (yes, we can lazily evaluate the index as well, but I would like to think about it separately). Once https://github.com/modin-project/modin/pull/4964 has been merged, there should no longer be an additional overhead.

Reply:
@anmyachev I'm a little confused about how this PR is helping. Here's my understanding. Right now we have two main cases for the parquet index:

1. index consists of one or more RangeIndex only: we block on getting index from metadata rather than reading whole file: https://github.com/modin-project/modin/blob/3ee4fa02ee69272aebf4fea005fe77268b166fbc/modin/core/io/column_stores/parquet_dispatcher.py#L521-L524
  I expect that to be cheap anyway.
1. index has at least one index that's not RangeIndex: we block on materializing the entire index from the partitions: https://github.com/modin-project/modin/blob/3ee4fa02ee69272aebf4fea005fe77268b166fbc/modin/core/io/column_stores/parquet_dispatcher.py#L528-L531

Since even after this PR we will continue blocking on getting the whole index in case (2), I don't see how this PR is helping.